---
title: "R Notebook"
output: Obtain AR regulon/target genes from TCGA PRAD and mCRPC data
---

#Load libraries
```{r}
library(RTN)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
```

#Downlaod TCGA PRAD RNAseq data from GDC
```{r}

#Download PRAD RNAseq data
query <- GDCquery(project = "TCGA-PRAD",
                  data.category = "Transcriptome Profiling",
                  data.type = "Gene Expression Quantification", 
                  workflow.type = "HTSeq - FPKM-UQ",
                  sample.type = c("Primary Tumor"))

GDCdownload(query)
tcgaBRCA_mRNA_data <- GDCprepare(query)


# Change column names in gene annotation for better summarizations
colnames(rowData(tcgaBRCA_mRNA_data)) <- c("ENSEMBL", "SYMBOL", "OG_ENSEMBL")


# Save the preprocessed data for subsequent analyses
save(tcgaBRCA_mRNA_data, file = "/Users/ds/Desktop/scratch/RNA_decon/data/tcgaPRAD_mRNA_data_preprocessed.RData")


```

#Read in data
```{r}
#Clinical anno
clin <- read.table("/Users/ds/Desktop/scratch/txt/202009_deepRNAseq_sample_full.txt", header = T) %>% dplyr::select("sample_id", "AR", "AR_enhancer")

#RNAseq CPM expression for mCRPC
rna <- read.table("/Users/ds/Desktop/scratch/RNA_decon/txt/mCRPC.CPM.normcounts.txt", header = T)
rna <- rna[!duplicated(rna$Genename),]
row.names(rna) <- rna$Genename
rna <- as.matrix((rna[,-1]))


#AR regulon and target genes
rtni <- readRDS("/Users/ds/Desktop/scratch/RNA_decon/data/rtni.rds")
```

#GSEA to obtain AR score
```{r}
#Summary of AR regulon 
tni.regulon.summary(rtni, regulatoryElements = "AR")

#View df of AR targets based on score
regulons <- tni.get(rtni, what = "regulons.and.mode")
regulons$AR %>% data.frame(Cor = .) %>% mutate(Genes = row.names(.)) %>% arrange(-Cor) %>% View()

#Run 2 tailed GSEA and DE analysis on mCRPC data 
rtni1st <- tni.gsea2(rtni, regulatoryElements = "AR")

#Get dES score i.e AR score for each sample
ar_activity <- tni.get(rtni1st, what = "regulonActivity")


#Plot heatmap comparing AR score with 
ar_activity <-  ar_activity$differential %>% data.frame(sample_id = gsub("\\.", "-", row.names(.))) %>% inner_join(y = clin, by = c("sample_id" = "sample_id")) %>% na.omit()


row.names(ar_activity) <- ar_activity$sample_id
colnames(ar_activity) <- c("AR_regulon_activity","sample_id" , "AR_gene_gof", "AR_enhancer_td")


#pheatmap(mat = t(ar_activity[,"AR_score", drop = F]), cluster_rows  = F, annotation_col = ar_activity[,c("AR_gene_gof", "AR_enhancer_td"), drop = F], fontsize_col = 7, cellheight = 10, main = "AR regulon activity in mCRPC")

#Write regulon list to txt file
#regulons$AR %>% names() %>% write.table("/Users/ds/Desktop/scratch/RNA_decon/txt/AR_regulon_mCPRC.txt", quote = F, sep = "\t", row.names = F, col.names = F)


ar_activity <- rna["AR",] %>% log1p() %>% data.frame(AR_RNA = ., sample_id = gsub("\\.", "-",names(.))) %>% merge(ar_activity)

saveRDS(object= ar_activity, file = "/Users/ds/Desktop/scratch/DNAm/Signature/AR_signature/AR_activity.rds")

```

